### DISCUSSION TITLE
Thoughts around supporting Node.js subpath imports

### DISCUSSION THREAD

**Discussion by thetutlage** [TypeStrong/ts-node#1799] [General]
The [subpath imports](https://nodejs.org/api/packages.html#imports) are the aliases one can use with the project to avoid typing relative paths during imports.

## Back story

- So, instead of typing:
  ```ts
  import foo from './src/foo.js'
  ```

- You can define a subpath import alias inside the `package.json` file
  ```json
  {
    "imports": {
      "#src/*": "./src/*.js"
    }
  }
  ```
- And then write imports as follows
  ```ts
  import foo from '#src/foo'
  ```

## The `tsc` behavior

Now there are a few issues in order to make it work.

- First, once I compile my TypeScript code to JavaScript, it is written inside a sub-directory called `build` and the subpath import alias becomes invalid. To overcome this issue, I define the import alias to look inside the `build` directory instead.
 
  ```json
  {
    "imports": {
      "#src/*": "./build/src/*.js"
    }
  }
  ```

- Luckily with Typescript >= 4.7, I have been able to compile this code using `tsc`. Even though the `build` directory does not exists pre-build, the TypeScript compiler is smartly able to resolve these imports.

## The `ts-node` behavior

However, now `ts-node` is not able to find the imports where the alias is pointing inside the `./build` directory (because there is no build directory). 

Now, the main question is, are there any on-going discussions or ideas on how to make subpath imports work when they are pointing within the `build` directory, because `tsc` is able to handle them.

I am using `node --loader=ts-node/esm` to execute the file

**Comment by cspotcode** [comment]
I believe that this issue can be renamed to something like this:
Support resolving from outDir to srcDir

Because that's really what we're talking about.  We do support subpath
imports.  The problem is that we don't attempt to jump from the outDir back
to the srcDir when resolving modules.

Does this assessment sound accurate to you?

**Reply by thetutlage** [reply]
I am not deeply aware of the features `ts-node` supports. So apologies for that :)

> Does this assessment sound accurate to you?

Yup, seems like a better explanation

**Reply by cspotcode** [reply]
No problem, I ask to determine the work that would be required to enable what you want to do.

package.json has a field `"publishConfig"` which can be used to overwrite package.json fields for publication.  This allows you to use different values in local development and when publishing.  I haven't tested this, but you might try using `"publishConfig"` to have two different `"imports"` objects: one for local development, pointing to your `src` directory, and another for publication, pointing to your `build` directory.

I haven't tested this so I will not be able to provide further guidance.  But if you research this option, you might consider sharing your results here to help other users of ts-node.  Or perhaps someone else has already done this, in which case you might find their posts by searching around.  I'm not sure.

**Reply by thetutlage** [reply]
I didn't find anything related to imports in relation with `publishConfig`. But yes, I can tweak my workflow to make it work somehow. 

However, I always wanted to understand if this is something you guys think worth supporting or addressing within the tool or not. If its not, then I am happy tweaking my workflow

**Reply by cspotcode** [reply]
Yeah, I've thought about it, and it's planned.  I think it is mentioned in the pinned "Epics" issue, somewhere.  Probably in the one about resolver.

**Reply by thetutlage** [reply]
Cool. Will look forward to it. Thanks for your time :)
